File: relocation_jobs/static/js/board.js
/** Board load path: paginated catalog from server, local UI filters only. */

import { state } from "./state.js";
import { $, beginTopLoadingProgress, finishLoadingProgress, setLoadingProgress } from "./utils.js";
import { fetchBoard, fetchBoardUserStats } from "./api.js";
import { releaseCompanyOrder } from "./render.js";
import { syncBoardView } from "./board-view.js";
import { beginScreenLoad, endScreenLoad, setScreenLoadProgress } from "./screen-loader.js";

function overlayLabel(options, page, requestChanged) {
  if (options.overlayLabel) return options.overlayLabel;
  if (requestChanged) return "Updating board…";
  if (page > 1) return "Loading page…";
  return "Loading board…";
}

function catalogScopeKey() {
  const country = $("country")?.value || "all";
  const ats = $("ats")?.value || "all";
  const location = $("location")?.value || "all";
  return `${country}|${ats}|${location}`;
}

function boardRequestKey() {
  const search = $("search")?.value.trim() || "";
  const filterIds = [
    "visaOnly",
    "hideApplied",
    "hideEmpty",
    "notAppliedOnly",
    "hidePositionApplied",
    "hidePositionRejected",
    "positionAppliedOnly",
    "positionRejectedOnly",
    "positionLookingToApplyOnly",
    "fetchOkOnly",
    "fetchProblemOnly",
  ];
  const flags = filterIds.map((id) => `${id}:${Boolean($(id)?.checked)}`).join("|");
  const sort = $("sortSelect")?.value === "name" ? "name" : "newest";
  return `${catalogScopeKey()}|${flags}|${search}|sort:${sort}`;
}

function boardPageSize() {
  return state.boardMeta?.page_size ?? 25;
}

export function boardTotalPages() {
  if (state.boardMeta?.total_pages != null) {
    return state.boardMeta.total_pages;
  }
  const total = state.boardMeta?.total_companies;
  if (total == null) return 1;
  return Math.max(1, Math.ceil(total / boardPageSize()));
}

function persistScopeSelection() {
  const country = $("country")?.value;
  if (country) localStorage.setItem("panel_country", country);
  if ($("ats")) localStorage.setItem("panel_ats", $("ats").value);
  if ($("location")) localStorage.setItem("panel_location", $("location").value);
}

export function showJobsLoading() {
  syncBoardView({ loading: true });
}

export function applyBoardView({ enterAnimation = false } = {}) {
  state.allCompanies = state.boardCatalog;
  const jobsEl = document.getElementById("jobs");
  if (enterAnimation) {
    jobsEl?.classList.add("board-enter");
  } else {
    jobsEl?.classList.remove("board-enter");
  }
  syncBoardView();
  if (enterAnimation && jobsEl) {
    window.clearTimeout(state.boardEnterTimer);
    state.boardEnterTimer = window.setTimeout(() => {
      jobsEl.classList.remove("board-enter");
      state.boardEnterTimer = null;
    }, 400);
  }
}

async function loadBoardCatalog(options = {}) {
  persistScopeSelection();
  const page = options.page ?? state.boardPage ?? 1;
  const requestKey = boardRequestKey();
  const preserveMeta = !options.requestChanged && state.boardRequestKey === requestKey;
  const data = await fetchBoard({ ...options, page });
  const prevTotal = preserveMeta ? state.boardMeta?.total_companies : null;
  const prevPageSize = preserveMeta ? state.boardMeta?.page_size : null;
  const prevTotalPages = preserveMeta ? state.boardMeta?.total_pages : null;
  state.boardCatalog = data.companies || [];
  state.boardMeta = preserveMeta
    ? { ...state.boardMeta, ...(data.meta || {}) }
    : { ...(data.meta || {}) };
  if (data.meta?.total_companies == null && prevTotal != null) {
    state.boardMeta.total_companies = prevTotal;
  }
  if (data.meta?.page_size == null && prevPageSize != null) {
    state.boardMeta.page_size = prevPageSize;
  }
  if (data.meta?.total_pages == null && prevTotalPages != null) {
    state.boardMeta.total_pages = prevTotalPages;
  }
  state.boardUserStats = data.user_stats || {};
  state.boardPage = data.meta?.page ?? page;
  state.boardScopeKey = catalogScopeKey();
  state.boardRequestKey = requestKey;
}

async function loadBoardUserStats() {
  state.boardUserStats = await fetchBoardUserStats();
}

export async function loadBoard(options = {}) {
  const requestChanged = state.boardRequestKey !== boardRequestKey();
  let page = options.page ?? state.boardPage ?? 1;
  if (requestChanged || (options.force && options.page == null)) {
    page = 1;
  }

  const preserveContent = options.preserveContent === true;
  const useOverlay = options.noOverlay !== true;
  const useTopBar = !useOverlay;
  const enterAnimation = options.enterAnimation !== false && !preserveContent;
  if (useOverlay) {
    beginScreenLoad(overlayLabel(options, page, requestChanged));
  } else if (useTopBar) {
    beginTopLoadingProgress(12);
  }

  releaseCompanyOrder();
  if (preserveContent) {
    syncBoardView({ loading: true, preserveContent: true });
  } else {
    showJobsLoading();
  }

  try {
    if (useOverlay) setScreenLoadProgress(20);
    else if (useTopBar) setLoadingProgress(28);
    await loadBoardCatalog({ ...options, page, requestChanged });
    if (useOverlay) setScreenLoadProgress(72);
    else if (useTopBar) setLoadingProgress(82);

    const totalPages = boardTotalPages();
    if (state.boardPage > totalPages) {
      await loadBoardCatalog({ ...options, page: 1, requestChanged: false });
    }

    if (options.refreshUserStats) {
      if (useOverlay) setScreenLoadProgress(86);
      else if (useTopBar) setLoadingProgress(90);
      await loadBoardUserStats();
    }

    if (useOverlay) setScreenLoadProgress(94);
    else if (useTopBar) setLoadingProgress(96);
    applyBoardView({ enterAnimation });
  } finally {
    if (useOverlay) endScreenLoad();
    else if (useTopBar) finishLoadingProgress();
  }
}

export async function goToBoardPage(page) {
  if (state.boardRequestKey !== boardRequestKey()) {
    await loadBoard({ overlayLabel: "Updating board…" });
    return;
  }
  const totalPages = boardTotalPages();
  const next = Math.max(1, Math.min(Number(page) || 1, totalPages));
  if (next === state.boardPage && state.boardCatalog.length) return;
  await loadBoard({ page: next, overlayLabel: "Loading page…" });
  document.getElementById("board-toolbar")?.scrollIntoView({ behavior: "smooth", block: "start" });
}

export async function refreshBoardUserStats() {
  await loadBoardUserStats();
  applyBoardView();
}

export { catalogScopeKey, boardRequestKey };

File: relocation_jobs/web/routes/board.py
from __future__ import annotations

import math

from flask import g, jsonify, request

from relocation_jobs.core.auth import login_required
from relocation_jobs.panel.board import (
    DEFAULT_BOARD_PAGE_SIZE,
    MAX_BOARD_PAGE_SIZE,
    load_catalog_board_page,
)
from relocation_jobs.panel.stats import compute_user_board_stats
from relocation_jobs.web.query import catalog_scope_flags, query_flags


def _latest_fetch_new_jobs(file_meta: list[dict]) -> int:
    return sum(int(row.get("last_fetch_new_jobs") or 0) for row in file_meta)


def _panel_flags() -> dict:
    flags = query_flags()
    return {
        "visa_only": flags["visa_only"],
        "hide_applied": flags["hide_applied"],
        "hide_empty": flags["hide_empty"],
        "not_applied_only": flags["not_applied_only"],
        "hide_position_applied": flags["hide_position_applied"],
        "hide_position_rejected": flags["hide_position_rejected"],
        "position_applied_only": flags["position_applied_only"],
        "position_rejected_only": flags["position_rejected_only"],
        "position_looking_to_apply_only": flags["position_looking_to_apply_only"],
        "fetch_ok_only": flags["fetch_ok_only"],
        "fetch_problem_only": flags["fetch_problem_only"],
    }


def register(app):
    @app.get("/api/board")
    @login_required
    def api_board():
        scope = catalog_scope_flags()
        timezone_name = (request.args.get("timezone") or "").strip() or None
        search = (request.args.get("q") or "").strip() or None
        page = max(request.args.get("page", 1, type=int) or 1, 1)
        page_size = request.args.get("page_size", DEFAULT_BOARD_PAGE_SIZE, type=int) or DEFAULT_BOARD_PAGE_SIZE
        page_size = max(1, min(page_size, MAX_BOARD_PAGE_SIZE))
        visible_offset = (page - 1) * page_size

        companies, file_meta, fetch_problem_count, total_visible, has_more = load_catalog_board_page(
            scope["country_key"],
            ats_type=scope["ats_type"],
            location=scope["location"],
            user_id=g.user_id,
            visible_offset=visible_offset,
            limit=page_size,
            search=search,
            panel_flags=_panel_flags(),
            count_total=(page == 1),
        )
        latest_fetch_new_jobs = _latest_fetch_new_jobs(file_meta)
        total_pages = None
        if total_visible is not None:
            total_pages = max(1, math.ceil(total_visible / page_size))
        return jsonify({
            "companies": companies,
            "meta": {
                "country": request.args.get("country", "all"),
                "ats_type": scope["ats_type"],
                "location": scope["location"],
                "fetch_problem_total": fetch_problem_count,
                "latest_fetch_new_jobs": latest_fetch_new_jobs,
                "page": page,
                "page_size": page_size,
                "total_companies": total_visible,
                "total_pages": total_pages,
                "has_more": has_more,
            },
            "user_stats": compute_user_board_stats(
                user_id=g.user_id,
                country_key=scope["country_key"],
                timezone_name=timezone_name,
                latest_fetch_new_jobs=latest_fetch_new_jobs,
            ),
        })

    @app.get("/api/board/stats")
    @login_required
    def api_board_stats():
        scope = catalog_scope_flags()
        timezone_name = (request.args.get("timezone") or "").strip() or None
        latest_fetch_new_jobs = request.args.get("latest_fetch_new_jobs", type=int) or 0
        return jsonify(compute_user_board_stats(
            user_id=g.user_id,
            country_key=scope["country_key"],
            timezone_name=timezone_name,
            latest_fetch_new_jobs=latest_fetch_new_jobs,
        ))
